topology/dynamic.go

```diff
@@ -0,0 +1,208 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package topology
+
+import (
+	"errors"
+	"math"
+	"sync"
+	"time"
+
+	"github.com/m3db/m3cluster/services"
+	"github.com/m3db/m3db/sharding"
+	"github.com/m3db/m3x/log"
+	"github.com/m3db/m3x/retry"
+	"github.com/m3db/m3x/watch"
+)
+
+var (
+	errNoConfigServiceClient      = errors.New("no config service client")
+	errInvalidConfigServiceClient = errors.New("invalid config service client")
+	errTimeOutInitializing        = errors.New("timed out initializing source")
+	errInvalidTopology            = errors.New("could not parse latest topology update")
+)
+
+type dynamicTopology struct {
+	sync.Mutex
+
+	s      xwatch.Source
+	w      xwatch.Watch
+	closed bool
+	logger xlog.Logger
+}
+
+func newDynamicTopology(options Options) (Topology, error) {
```

Comment:
Might be worth moving dynamic topology and static topology into their own packages `static` and `dynamic` under `topology` package as it's kind of weird they share the same options because depending which one you're using you may or may not have to set some properties.  Thoughts?